---
title: "Applying STANCE to Spatial Transcriptomics Data"
author: "Haohao Su"
date: "2024-06-29"
output: html_document
---

```{r setup, message=FALSE}
library(STANCE)
library(Matrix)
library(ggplot2)
library(reshape2)
library(dplyr)
```

## Load in data
[Mouse olfactory bulb replicate No.8](https://www.science.org/doi/10.1126/science.aaf2403) is a $100 \mu m^2$ resolution spatial transcriptomics dataset of 15,928 genes measured on 262 spots. Following the procedure of [Miller et al. (2022)](https://www.nature.com/articles/s41467-022-30033-z), data preprocessing and cell type deconvolution by [STdeconvolve](https://jef.works/STdeconvolve/) have been conducted, resulting in the **mOB** dataset, which is a list containing the gene expression, spot location coordinates and cell type compositions. 

Specifically, **counts** contains the gene expression of 7,365 genes across 260 spots.
```{r}
data(mOB)
mOB$counts[1:5,1:5]
```

**pos** is the spatial location coordinates of the 260 spots.
```{r}
mOB$pos[1:5,]
```

**prop** is cell type proportion of 12 cell types across the 260 spots.
```{r}
mOB$prop[1:5,]
```

## Create STANCE object
To run STANCE, we need to first create a STANCE object using `creatSTANCEobject` function, which requires 4 inputs:
1. **counts**: gene expression matrix with genes in rows and spots in columns;
2. **pos**: dimensional location coordinates matrix. The rownames of **pos** matrix should match the colnames of **counts** matrix;
3. **prop**: he cell type proportion of each spot. Rows represent spots and columns represent cell types. The rownames of **prop** matrix should match the colnames of **counts** matrix;
4. **covariates** (default NULL) the covariate (if any) design matrix modeling gene expression. The rownames of **covariates** matrix should match the colnames of **counts** matrix.

All the inputs should be convertible to 'matrix' form.
```{r}
mySTANCE <- creatSTANCEobject(counts = mOB$counts,
                       pos = mOB$pos,
                       prop = mOB$prop,
                       covariates = NULL)
```

## Preprocess the input data
Normalize gene expression count data, scale location coordinates and perform quality conrol steps, including removing spots with gene expression less than $10$, and removing genes whose proportion of zero-express spots greater than $0.05$ by default. `normalized = TRUE` is used to specify that the gene expression input has been normalized so as to skip the normalization step.
```{r}
mySTANCE <- data_preprocess(object = mySTANCE, 
                            gene.threshold = 0.05, spot.threshold = 10,
                            normalized = FALSE)
```

## Build Gaussian kernel matrix
Build the Gaussian kernel matrix to model spatial correlation between spots. The Gaussian kernel bandwidth will be selected automatically:
- for datasets with 5000 spots or fewer, non-parametric Sheather-Jones' bandwidths are computed for each gene, and the median of these values is used as the common bandwidth; 
- for datasets with more than 5000 spots, Silverman's bandwidths are computed for each gene, and the median of these gene-specific bandwidths is used as the common bandwidth.
```{r}
mySTANCE <- build_kernelMatrix(object = mySTANCE)
```

## Obtain lists for important model matrices
This step is to construct the important matrices, including design matrices $\mathbf{X}$, covariance matrices (i.e. $\mathbf{K}$ for spatial effects $\gamma_k$ and $\mathbf{I}$ for residual error $\varepsilon$) and $\mathbf{\Sigma}_k = \mathbf{\Pi}_k \mathbf{K} \mathbf{\Pi}_k^T$ matrices for all the $\gamma_k$'s, before fitting STANCE model.
```{r}
mySTANCE <- get_modelMatrices(object = mySTANCE)
```

## Perform STANCE overall test for SVGs & ctSVGs
Here, for effiency, we skip the bias correction procedure. Use "Benjamini-Yekutieli" method to adjust p-values to control false discovery rate.
```{r}
mySTANCE <- runTest1(object = mySTANCE, 
                     correction = F, pv.adjust = "BY")
```

The function outputs a data frame containing the test statistic values, the associated p-values as well as the "BY" adjusted p-values.
```{r}
head(mySTANCE@Test_1)
```

## Perform STANCE individual test 
```{r}
# List for all the genes in the experiment
genes.list <- rownames(mySTANCE@gene_expression)
# List for SVGs & ctSVGs detected by STANCE overall test
genes.sig.1 <- genes.list[mySTANCE@Test_1$p_value_adj < 0.05]
```

Perform STANCE individual test on each of genes passing the overall test (we called these genes "**utSVGs**"). `Cell_types_to_test = NULL` indicates that all the cell types would be tested. `Genes_to_test = genes.sig.1[1:100]` specifies that the first 100 significant genes detected by overall test will be tested.
```{r}
mySTANCE <- runTest2(object = mySTANCE, 
                     Genes_to_test = genes.sig.1[1:100], Cell_types_to_test = NULL, 
                     pv.adjust = "BY")
```

The function outputs a list containing a data frame of test statistic values, as well as the associated p-values for each cell type.
```{r}
# Individual testing results for cell type "1"
head(mySTANCE@Test_2[["1"]])
# Individual testing results for cell type "2"
head(mySTANCE@Test_2[["2"]])
```

## Estimate the variance components of STANCE model for all the genes
`Genes_to_test = genes.sig.1[1:100]` specifies that the first 100 significant genes detected by overall test will be estimated.
```{r}
mySTANCE <- varcomp_est(object = mySTANCE, Genes_to_est = genes.sig.1[1:100])
```

```{r}
head(mySTANCE@VarComp_estimates)
```

## Display top ctSVGs for the cell type of interest
Pick out top 20 ctSVGs (if less than 20, then all the ctSVGs will be displayed) for cell type of interest (`CT_of_interest = "1"`) .
```{r}
mySTANCE <- CT_topGenes(object = mySTANCE, 
                        CT_of_interest = "1", numTopGenes = 20)
# Print the list of top genes for cell type "1"
cat(paste0('The top cell type specific SVGs for cell type \"',
           mySTANCE@cell_type_top_genes$cell_type_of_interest,
           '\": \n',
           paste(mySTANCE@cell_type_top_genes$top_genes, collapse = ","),
           '. \n'))
```

Draw a stacked bar plot for top genes visualizing the proportion of variance explained by all the cell types and error.
```{r}
print(mySTANCE@cell_type_top_genes$stacked_bar_plot)
```
